lib/command.js

```diff
@@ -1189,12 +1221,18 @@ Expecting one of '${allowedValues.join("', '")}'`);
 
   _chainOrCall(promise, fn) {
     // thenable
-    if (promise && promise.then && typeof promise.then === 'function') {
+    if (isThenable(promise)) {
       // already have a promise, chain callback
       return promise.then(() => fn());
     }
+
     // callback might return a promise
-    return fn();
+    const result = fn();
+    if (!PRODUCTION && !this._asyncParsing && isThenable(result)) {
```

Comment:
Why not have this test in `.parse()` itself? Save having a global as well as have the code right where it is relevant.

Reply:
`_chainOrCall()` could later be used not only for the chain returned from `_parseCommand()`, see #1915 ([comment](https://github.com/tj/commander.js/pull/1915#discussion_r1277289430)).

Reply:
Actually, it might be a good idea to show a different warning for argParsers and wrap the relevant code in a new method called `_chainOrCallArgParser()`.

Reply:
Oh, actually, I see now the right place for the argParser warning in #1915 is `_parseArg()`.

Reply:
Resolved in https://github.com/tj/commander.js/pull/1917/commits/3572657066e281b98eb5a653d00b26de7ceee958 (`PRODUCTION` still has to be global, though).

Reply:
The "global" I was referring to was `_asyncParsing`. I should have said an instance property. It is still in the code, but I think no longer being used after latest changes.

Reply:
Oh yes, that's right. Fixed in https://github.com/tj/commander.js/pull/1917/commits/adde7ffcf0a4b02bbcef50dc7b8dc5b4261b490d.